README.md
# dfply

### Version: 0.3.2

> Note: Version 0.3.0 is the first big update in awhile, and changes a lot of
the "base" code. The `pandas-ply` package is no longer being imported. I have coded
my own version of the "symbolic" objects that I was borrowing from `pandas-ply`. Also,
I am no longer supporting Python 2, sorry!

> **In v0.3 `groupby` has been renamed to `group_by` to mirror the `dplyr` function.
If this breaks your legacy code, one possible fix is to have `from dfply.group import group_by as groupby`
in your package imports.**

The `dfply` package makes it possible to do R's `dplyr`-style data manipulation with pipes
in python on pandas DataFrames.

This is an alternative to [`pandas-ply`](https://github.com/coursera/pandas-ply)
and [`dplython`](https://github.com/dodger487/dplython), which both engineer `dplyr`
syntax and functionality in python. There are probably more packages that attempt
to enable `dplyr`-style dataframe manipulation in python, but those are the two I
am aware of.

`dfply` uses a decorator-based architecture for the piping functionality and
to "categorize" the types of data manipulation functions. The goal of this  
architecture is to make `dfply` concise and easily extensible, simply by chaining
together different decorators that each have a distinct effect on the wrapped
function. There is a more in-depth overview of the decorators and how `dfply` can be
customized below.

`dfply` is intended to mimic the functionality of `dplyr`. The syntax
is the same for the most part, but will vary in some cases as Python is a
considerably different programming language than R.

A good amount of the core functionality of `dplyr` is complete, and the remainder is
actively being added in. Going forward I hope functionality that is not
directly part of `dplyr` to be incorporated into `dfply` as well. This is not
intended to be an absolute mimic of `dplyr`, but instead a port of the _ease,
convenience and readability_ the `dplyr` package provides for data manipulation
tasks.

**Expect frequent updates to the package version as features are added and
any bugs are fixed.**

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Overview of functions](#overview-of-functions)
  - [The `>>` and `>>=` pipe operators](#the--and--pipe-operators)
  - [The `X` DataFrame symbol](#the-x-dataframe-symbol)
  - [Selecting and dropping](#selecting-and-dropping)
    - [`select()` and `drop()` functions](#select-and-drop-functions)
    - [Selection using the inversion `~` operator on symbolic columns](#selection-using-the-inversion--operator-on-symbolic-columns)
    - [Selection filter functions](#selection-filter-functions)
  - [Subsetting and filtering](#subsetting-and-filtering)
    - [`row_slice()`](#row_slice)
    - [`sample()`](#sample)
    - [`distinct()`](#distinct)
    - [`mask()`](#mask)
  - [DataFrame transformation](#dataframe-transformation)
    - [`mutate()`](#mutate)
    - [`transmute()`](#transmute)
  - [Grouping](#grouping)
    - [`group_by()` and `ungroup()`](#group_by-and-ungroup)
  - [Reshaping](#reshaping)
    - [`arrange()`](#arrange)
    - [`rename()`](#rename)
    - [`gather()`](#gather)
    - [`spread()`](#spread)
    - [`separate()`](#separate)
    - [`unite()`](#unite)
  - [Joining](#joining)
    - [`inner_join()`](#inner_join)
    - [`outer_join()` or `full_join()`](#outer_join-or-full_join)
    - [`left_join()`](#left_join)
    - [`right_join()`](#right_join)
    - [`semi_join()`](#semi_join)
    - [`anti_join()`](#anti_join)
  - [Set operations](#set-operations)
    - [`union()`](#union)
    - [`intersect()`](#intersect)
    - [`set_diff()`](#set_diff)
  - [Binding](#binding)
    - [`bind_rows()`](#bind_rows)
    - [`bind_cols()`](#bind_cols)
  - [Summarization](#summarization)
    - [`summarize()`](#summarize)
    - [`summarize_each()`](#summarize_each)
- [Embedded column functions](#embedded-column-functions)
  - [Window functions](#window-functions)
    - [`lead()` and `lag()`](#lead-and-lag)
    - [`between()`](#between)
    - [`dense_rank()`](#dense_rank)
    - [`min_rank()`](#min_rank)
    - [`cumsum()`](#cumsum)
    - [`cummean()`](#cummean)
    - [`cummax()`](#cummax)
    - [`cummin()`](#cummin)
    - [`cumprod()`](#cumprod)
  - [Summary functions](#summary-functions)
    - [`mean()`](#mean)
    - [`first()`](#first)
    - [`last()`](#last)
    - [`nth()`](#nth)
    - [`n()`](#n)
    - [`n_distinct()`](#n_distinct)
    - [`IQR()`](#iqr)
    - [`colmin()`](#colmin)
    - [`colmax()`](#colmax)
    - [`median()`](#median)
    - [`var()`](#var)
    - [`sd()`](#sd)
- [Extending `dfply` with custom functions](#extending-dfply-with-custom-functions)
  - [Case 1: A custom "pipe" function with `@dfpipe`](#case-1-a-custom-pipe-function-with-dfpipe)
  - [Case 2: A function that works with symbolic objects using `@make_symbolic`](#case-2-a-function-that-works-with-symbolic-objects-using-make_symbolic)
    - [Without symbolic arguments, `@make_symbolic` functions work like normal functions!](#without-symbolic-arguments-make_symbolic-functions-work-like-normal-functions)
- [Advanced: understanding base `dfply` decorators](#advanced-understanding-base-dfply-decorators)
  - [The `Intention` class](#the-intention-class)
  - [`@pipe`](#pipe)
  - [`@group_delegation`](#group_delegation)
  - [`@symbolic_evaluation`](#symbolic_evaluation)
    - [Controlling `@symbolic_evaluation` with the `eval_symbols` argument](#controlling-symbolic_evaluation-with-the-eval_symbols-argument)
  - [`@dfpipe`](#dfpipe)
  - [`@make_symbolic`](#make_symbolic)
- [Contributing](#contributing)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview of functions

### The `>>` and `>>=` pipe operators

dfply works directly on pandas DataFrames, chaining operations on the data with
the `>>` operator, or alternatively starting with `>>=` for inplace operations.

```python
from dfply import *

diamonds >> head(3)

   carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
```

You can chain piped operations, and of course assign the output to a new
DataFrame.

```python
lowprice = diamonds >> head(10) >> tail(3)

lowprice

   carat        cut color clarity  depth  table  price     x     y     z
7   0.26  Very Good     H     SI1   61.9   55.0    337  4.07  4.11  2.53
8   0.22       Fair     E     VS2   65.1   61.0    337  3.87  3.78  2.49
9   0.23  Very Good     H     VS1   59.4   61.0    338  4.00  4.05  2.39
```

Inplace operations are done with the first pipe as `>>=` and subsequent pipes
as `>>`.

```python
diamonds >>= head(10) >> tail(3)

diamonds

   carat        cut color clarity  depth  table  price     x     y     z
7   0.26  Very Good     H     SI1   61.9   55.0    337  4.07  4.11  2.53
8   0.22       Fair     E     VS2   65.1   61.0    337  3.87  3.78  2.49
9   0.23  Very Good     H     VS1   59.4   61.0    338  4.00  4.05  2.39
```

When using the inplace pipe, the DataFrame is not required on the left hand
side of the `>>=` pipe and the DataFrame variable is overwritten with the
output of the operations.


### The `X` DataFrame symbol

The DataFrame as it is passed through the piping operations is represented
by the symbol `X`. It records the actions you want to take (represented by
the `Intention` class), but does not evaluate them until the appropriate time.
Operations on the DataFrame are deferred. Selecting
two of the columns, for example, can be done using the symbolic `X` DataFrame
during the piping operations.

```python
diamonds >> select(X.carat, X.cut) >> head(3)

   carat      cut
0   0.23    Ideal
1   0.21  Premium
2   0.23     Good
```


### Selecting and dropping

#### `select()` and `drop()` functions

There are two functions for selection, inverse of each other: `select` and
`drop`. The `select` and `drop` functions accept string labels, integer positions,
and/or symbolically represented column names (`X.column`). They also accept symbolic "selection
filter" functions, which will be covered shortly.

The example below selects "cut", "price", "x", and "y" from the diamonds dataset.

```python
diamonds >> select(1, X.price, ['x', 'y']) >> head(2)

       cut  price     x     y
0    Ideal    326  3.95  3.98
1  Premium    326  3.89  3.84
```

If you were instead to use `drop`, you would get back all columns besides those specified.

```python
diamonds >> drop(1, X.price, ['x', 'y']) >> head(2)

   carat color clarity  depth  table     z
0   0.23     E     SI2   61.5   55.0  2.43
1   0.21     E     SI1   59.8   61.0  2.31
```


#### Selection using the inversion `~` operator on symbolic columns

One particularly nice thing about `dplyr`'s selection functions is that you can
drop columns inside of a select statement by putting a subtraction sign in front,
like so: `... %>% select(-col)`. The same can be done in `dfply`, but instead of
the subtraction operator you use the tilde `~`.

For example, let's say I wanted to select any column _except_ carat, color, and
clarity in my dataframe. One way to do this is to specify those for removal using
the `~` operator like so:


```python
diamonds >> select(~X.carat, ~X.color, ~X.clarity) >> head(2)

       cut  depth  table  price     x     y     z
0    Ideal   61.5   55.0    326  3.95  3.98  2.43
1  Premium   59.8   61.0    326  3.89  3.84  2.31
```

Note that if you are going to use the inversion operator, you _must_ place it
prior to the symbolic `X` (or a symbolic such as a selection filter function, covered
next). For example, using the inversion operator on a list of columns will
result in an error:

```python
diamonds >> select(~[X.carat, X.color, X.clarity]) >> head(2)

TypeError: bad operand type for unary ~: 'list'
```


#### Selection filter functions

The vanilla `select` and `drop` functions are useful, but there are a variety of
selection functions inspired by `dplyr` available to make selecting and dropping
columns a breeze. These functions are intended to be put inside of the `select` and
`drop` functions, and can be paired with the `~` inverter.

First, a quick rundown of the available functions:
- `starts_with(prefix)`: find columns that start with a string prefix.
- `ends_with(suffix)`: find columns that end with a string suffix.
- `contains(substr)`: find columns that contain a substring in their name.
- `everything()`: all columns.
- `columns_between(start_col, end_col, inclusive=True)`: find columns between a specified start and end column.
The `inclusive` boolean keyword argument indicates whether the end column should be included or not.
- `columns_to(end_col, inclusive=True)`: get columns up to a specified end column. The `inclusive`
argument indicates whether the ending column should be included or not.
- `columns_from(start_col)`: get the columns starting at a specified column.

The selection filter functions are best explained by example. Let's say I wanted to
select only the columns that started with a "c":

```python
diamonds >> select(starts_with('c')) >> head(2)

   carat      cut color clarity
0   0.23    Ideal     E     SI2
1   0.21  Premium     E     SI1
```

The selection filter functions are instances of the class `Intention`, just like the
`X` placeholder, and so I can also use the inversion operator with them. For example,
I can alternatively select the columns that do not start with "c":

```python
diamonds >> select(~starts_with('c')) >> head(2)

   depth  table  price     x     y     z
0   61.5   55.0    326  3.95  3.98  2.43
1   59.8   61.0    326  3.89  3.84  2.31
```

They work the same inside the `drop` function, but with the intention of removal.
I could, for example, use the `columns_from` selection filter to drop all columns
from "price" onwards:

```python
diamonds >> drop(columns_from(X.price)) >> head(2)

   carat      cut color clarity  depth  table
0   0.23    Ideal     E     SI2   61.5   55.0
1   0.21  Premium     E     SI1   59.8   61.0
```

As the example above shows, you can use symbolic column names inside of the
selection filter function! You can also mix together selection filters and standard
selections inside of the same `select` or `drop` command.

For my next trick, I will select the first two columns, the last two columns, and
the "depth" column using a mixture of selection techniques:

```python
diamonds >> select(columns_to(1, inclusive=True), 'depth', columns_from(-2)) >> head(2)

   carat      cut  depth     y     z
0   0.23    Ideal   61.5  3.98  2.43
1   0.21  Premium   59.8  3.84  2.31
```


### Subsetting and filtering

#### `row_slice()`

Slices of rows can be selected with the `row_slice()` function. You can pass
single integer indices or a list of indices to select rows as with. This is
going to be the same as using pandas' `.iloc`.

```python
diamonds >> row_slice([10,15])

    carat      cut color clarity  depth  table  price     x     y     z
10   0.30     Good     J     SI1   64.0   55.0    339  4.25  4.28  2.73
15   0.32  Premium     E      I1   60.9   58.0    345  4.38  4.42  2.68
```

Note that this can also be used with the `group_by` function, and will operate
like a call to `.iloc` on each group. The `group_by` pipe function is
covered later, but it essentially works the same as pandas `.groupby` (with a
few subtle differences).

```python
diamonds >> group_by('cut') >> row_slice(5)

     carat        cut color clarity  depth  table  price     x     y     z
128   0.91       Fair     H     SI2   64.4   57.0   2763  6.11  6.09  3.93
20    0.30       Good     I     SI2   63.3   56.0    351  4.26  4.30  2.71
40    0.33      Ideal     I     SI2   61.2   56.0    403  4.49  4.50  2.75
26    0.24    Premium     I     VS1   62.5   57.0    355  3.97  3.94  2.47
21    0.23  Very Good     E     VS2   63.8   55.0    352  3.85  3.92  2.48
```



#### `sample()`

The `sample()` function functions exactly the same as pandas' `.sample()` method
for DataFrames. Arguments and keyword arguments will be passed through to the
DataFrame sample method.

```python
diamonds >> sample(frac=0.0001, replace=False)

       carat        cut color clarity  depth  table  price     x     y     z
19736   1.02      Ideal     E     VS1   62.2   54.0   8303  6.43  6.46  4.01
37159   0.32    Premium     D     VS2   60.3   60.0    972  4.44  4.42  2.67
1699    0.72  Very Good     E     VS2   63.8   57.0   3035  5.66  5.69  3.62
20955   1.71  Very Good     J     VS2   62.6   55.0   9170  7.58  7.65  4.77
5168    0.91  Very Good     E     SI2   63.0   56.0   3772  6.12  6.16  3.87


diamonds >> sample(n=3, replace=True)

       carat        cut color clarity  depth  table  price     x     y     z
52892   0.73  Very Good     G     SI1   60.6   59.0   2585  5.83  5.85  3.54
39454   0.57      Ideal     H     SI2   62.3   56.0   1077  5.31  5.28  3.30
39751   0.43      Ideal     H    VVS1   62.3   54.0   1094  4.84  4.85  3.02
```

#### `distinct()`

Selection of unique rows is done with `distinct()`, which similarly passes
arguments and keyword arguments through to the DataFrame's `.drop_duplicates()`
method.

```python
diamonds >> distinct(X.color)

    carat        cut color clarity  depth  table  price     x     y     z
0    0.23      Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
3    0.29    Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4    0.31       Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75
7    0.26  Very Good     H     SI1   61.9   55.0    337  4.07  4.11  2.53
12   0.22    Premium     F     SI1   60.4   61.0    342  3.88  3.84  2.33
25   0.23  Very Good     G    VVS2   60.4   58.0    354  3.97  4.01  2.41
28   0.23  Very Good     D     VS2   60.5   61.0    357  3.96  3.97  2.40
```


#### `mask()`

Filtering rows with logical criteria is done with `mask()`, which accepts
boolean arrays "masking out" False labeled rows and keeping True labeled rows.
These are best created with logical statements on symbolic Series objects as
shown below. Multiple criteria can be supplied as arguments and their intersection
will be used as the mask.

```python
diamonds >> mask(X.cut == 'Ideal') >> head(4)

    carat    cut color clarity  depth  table  price     x     y     z
0    0.23  Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
11   0.23  Ideal     J     VS1   62.8   56.0    340  3.93  3.90  2.46
13   0.31  Ideal     J     SI2   62.2   54.0    344  4.35  4.37  2.71
16   0.30  Ideal     I     SI2   62.0   54.0    348  4.31  4.34  2.68

diamonds >> mask(X.cut == 'Ideal', X.color == 'E', X.table < 55, X.price < 500)

       carat    cut color clarity  depth  table  price     x     y     z
26683   0.33  Ideal     E     SI2   62.2   54.0    427  4.44  4.46  2.77
32297   0.34  Ideal     E     SI2   62.4   54.0    454  4.49  4.52  2.81
40928   0.30  Ideal     E     SI1   61.6   54.0    499  4.32  4.35  2.67
50623   0.30  Ideal     E     SI2   62.1   54.0    401  4.32  4.35  2.69
50625   0.30  Ideal     E     SI2   62.0   54.0    401  4.33  4.35  2.69
```

Alternatively, `mask()` can also be called using the alias `filter_by()`:

```python
diamonds >> filter_by(X.cut == 'Ideal', X.color == 'E', X.table < 55, X.price < 500)

       carat    cut color clarity  depth  table  price     x     y     z
26683   0.33  Ideal     E     SI2   62.2   54.0    427  4.44  4.46  2.77
32297   0.34  Ideal     E     SI2   62.4   54.0    454  4.49  4.52  2.81
40928   0.30  Ideal     E     SI1   61.6   54.0    499  4.32  4.35  2.67
50623   0.30  Ideal     E     SI2   62.1   54.0    401  4.32  4.35  2.69
50625   0.30  Ideal     E     SI2   62.0   54.0    401  4.33  4.35  2.69
```

#### `pull()`

`pull` simply retrieves a column and returns it as a pandas series, in case you only care about one particular column at the end of your pipeline. Columns can be specified either by their name (string) or an integer.

The default returns the last column (on the assumption that's the column you've created most recently).

Example:

```python
(diamonds
 >> filter_by(X.cut == 'Ideal', X.color == 'E', X.table < 55, X.price < 500)
 >> pull('carat'))

 26683    0.33
 32297    0.34
 40928    0.30
 50623    0.30
 50625    0.30
 Name: carat, dtype: float64
```


### DataFrame transformation

#### `mutate()`

New variables can be created with the `mutate()` function (named that way to match
`dplyr`).

```python
diamonds >> mutate(x_plus_y=X.x + X.y) >> select(columns_from('x')) >> head(3)

      x     y     z  x_plus_y
0  3.95  3.98  2.43      7.93
1  3.89  3.84  2.31      7.73
2  4.05  4.07  2.31      8.12
```

Multiple variables can be created in a single call.

```python
diamonds >> mutate(x_plus_y=X.x + X.y, y_div_z=(X.y / X.z)) >> select(columns_from('x')) >> head(3)

      x     y     z  x_plus_y   y_div_z
0  3.95  3.98  2.43      7.93  1.637860
1  3.89  3.84  2.31      7.73  1.662338
2  4.05  4.07  2.31      8.12  1.761905
```

> Note: In Python the new variables created with mutate may not be guaranteed
to be created in the same order that they are input into the function call, though

RELEASES.txt
v0.3.3
- Hotfix for parsing left_on and right_on

TODO: Need to figure out fix to the inversion of symbol issue. Somewhat complicated.

v0.3.2
Various PRs added fixing bugs, etc.

v0.3.1
This update is almost solely the pull requests by @bleearmstrong that were sitting
in the repo waiting. There were some minor bug-fixes and changes too.

- `select_if` and `drop_if` are now available to perform selection according to a function
- `mutate_if` allows variable creation if a criterion function is evaluated as True
- `row_number` window function is available (same as rank(method='first'))
- `distinct` can take no arguments, making it equivalent to `drop_duplicates` with no arguments

v0.3.0
Lots and lots of big changes here. Total reworking of the internal functionality.
The good news is that it should (basically) work the same as before, but
hopefully better. Obviously keep in mind that this is still beta and there will
be plenty of bugs to work out on the horizon, but the preexisting tests pass for now...
There is not backward compatibility with old versions as the decorator names
have changed, but again, the functionality is otherwise the same.

Some major things (see readme for details):
- Moved entirely to python 3 support only. It may still work in python 2, but don't count on it.
- pandas-ply is no longer required; It was brittle and so rolled my own stuff.
- Selection "subfunctions" are now working and the selection functions have been changed in light of this.


v0.2.4
- Bug fixed in semi-join and anti-join
- top_n added

v0.2.3
Inplace piping added using the `>>=` operator. The `pipe` decorator internals
have been changed to make this possible through the addition of an `__rshift__`
implementation and chaining pipes together until evaluated against a
DataFrame.


v0.2.2
- Added docstrings to functions and classes.
- Added the `case_when` function.
- Fixed `arrange` to work with symbolic functions like `desc` in the function
  call.
- Added `cumany` and `cumall` window functions.
- Added `if_else` function.
- Added `na_if` function.
- Added `percent_rank` function.
- Reorganization of decorator functions (better subclassing).


v0.2.1
Fixed an issue with the way the decorators were structured, particularly
the @make_symbolic, that would cause problems with returning Call objects
that would not evaluate properly. Hopefully this is now resolved.

The "coalesce" function was added from dplyr.

Some code was moved from base.py to the new vector.py file. The vector.py
file now contains functions that specifically perform operations on
series or numpy arrays (such as coalesce, desc, etc.). Test files have been
reorganized accordingly.


v0.2.0
This release now introduces the @make_symbolic decorator, which can wrap
functions to allow them to evaluate later. This is particularly (and perhaps
only) useful when you embed functions as arguments to other functions. For
example the summary and window functions.

The code for the symbolic handling decorators has been reworked. They now
inherit from a common class since they shared patterns in their code.

- @make_symbolic decorator
- README updates
- desc() and order_series_by() functions
- re-imagining of the code for @symbolic_evaluation, @symbolic_reference
  (the functionality remains unchanged)
- window and summary functions, along with their tests, have been moved around
  to other files.


v0.1.10
- `separate` and `unite` functions added.
- Summary functions added for series operations.
- README improved dramatically.
- Function docstrings added to more functions (still not all).

v0.1.9
Moved unit tests into individual files that reflect the categories of the
functions/features they are testing. Some small bugs have been fixed as well.

v0.1.8
The pipe decorator now copies the dataframe upon each chained function, along
with the `_grouped_by` attribute, if any. Before, operations with the pipe
functions were modifying the original dataframe (such as `mutate`).

v0.1.7
Restructuring of package to include `diamonds.csv` with pip installation
and require `six` and `pandas-ply` rather than come pre-packaged with them.

v0.1.6
Added window functions:
`dense_rank`
`min_rank`
`cumsum`
`cummean`
`cummax`
`cummin`
`cumprod`


v0.1.5
dplyr set operations added thanks to bleearmstrong.

`df >> union(other)`
Rows that appear in either `df` or `other`.

`df >> intersect(other)`
Rows that appear in both `df` and `other`

`df >> set_diff(other)`
Rows that appear in `df` but not `other`.


dfply/reshape.py
from .base import *
import re


# ------------------------------------------------------------------------------
# Sorting
# ------------------------------------------------------------------------------

@dfpipe
def arrange(df, *args, **kwargs):
    """Calls `pandas.DataFrame.sort_values` to sort a DataFrame according to
    criteria.

    See:
    http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html

    For a list of specific keyword arguments for sort_values (which will be
    the same in arrange).

    Args:
        *args: Symbolic, string, integer or lists of those types indicating
            columns to sort the DataFrame by.

    Kwargs:
        **kwargs: Any keyword arguments will be passed through to the pandas
            `DataFrame.sort_values` function.
    """

    flat_args = [a for a in flatten(args)]

    series = [df[arg] if isinstance(arg, str) else
              df.iloc[:, arg] if isinstance(arg, int) else
              pd.Series(arg) for arg in flat_args]

    sorter = pd.concat(series, axis=1).reset_index(drop=True)
    sorter = sorter.sort_values(sorter.columns.tolist(), **kwargs)
    return df.iloc[sorter.index, :]


# ------------------------------------------------------------------------------
# Renaming
# ------------------------------------------------------------------------------

@pipe
@symbolic_evaluation(eval_as_label=True)
def rename(df, **kwargs):
    """Renames columns, where keyword argument values are the current names
    of columns and keys are the new names.

    Args:
        df (:obj:`pandas.DataFrame`): DataFrame passed in via `>>` pipe.

    Kwargs:
        **kwargs: key:value pairs where keys are new names for columns and
            values are current names of columns.
    """

    return df.rename(columns={v: k for k, v in kwargs.items()})


# ------------------------------------------------------------------------------
# Elongate
# ------------------------------------------------------------------------------

@pipe
@symbolic_evaluation(eval_as_label=['*'])
def gather(df, key, values, *args, **kwargs):
    """
    Melts the specified columns in your DataFrame into two key:value columns.

    Args:
        key (str): Name of identifier column.
        values (str): Name of column that will contain values for the key.
        *args (str, int, symbolic): Columns to "melt" into the new key and
            value columns. If no args are specified, all columns are melted
            into they key and value columns.

    Kwargs:
        add_id (bool): Boolean value indicating whether to add a `"_ID"`
            column that will preserve information about the original rows
            (useful for being able to re-widen the data later).

    Example:
        diamonds >> gather('variable', 'value', ['price', 'depth','x','y','z']) >> head(5)

           carat      cut color clarity  table variable  value
        0   0.23    Ideal     E     SI2   55.0    price  326.0
        1   0.21  Premium     E     SI1   61.0    price  326.0
        2   0.23     Good     E     VS1   65.0    price  327.0
        3   0.29  Premium     I     VS2   58.0    price  334.0
        4   0.31     Good     J     SI2   58.0    price  335.0
    """

    if len(args) == 0:
        args = df.columns.tolist()
    else:
        args = [a for a in flatten(args)]

    if kwargs.get('add_id', False):
        df = df.assign(_ID=np.arange(df.shape[0]))

    columns = df.columns.tolist()
    id_vars = [col for col in columns if col not in args]
    return pd.melt(df, id_vars, list(args), key, values)


# ------------------------------------------------------------------------------
# Widen
# ------------------------------------------------------------------------------

def convert_type(df, columns):
    """
    Helper function that attempts to convert columns into their appropriate
    data type.
    """
    # taken in part from the dplython package
    out_df = df.copy()
    for col in columns:
        column_values = pd.Series(out_df[col].unique())
        column_values = column_values[~column_values.isnull()]
        # empty
        if len(column_values) == 0:
            continue
        # boolean
        if set(column_values.values) < {'True', 'False'}:
            out_df[col] = out_df[col].map({'True': True, 'False': False})
            continue
        # numeric
        if pd.to_numeric(column_values, errors='coerce').isnull().sum() == 0:
            out_df[col] = pd.to_numeric(out_df[col], errors='ignore')
            continue
        # datetime
        if pd.to_datetime(column_values, errors='coerce').isnull().sum() == 0:
            out_df[col] = pd.to_datetime(out_df[col], errors='ignore',
                                         infer_datetime_format=True)
            continue

    return out_df


@pipe
@symbolic_evaluation(eval_as_label=['*'])
def spread(df, key, values, convert=False):
    """
    Transforms a "long" DataFrame into a "wide" format using a key and value
    column.

    If you have a mixed datatype column in your long-format DataFrame then the
    default behavior is for the spread columns to be of type `object`, or
    string. If you want to try to convert dtypes when spreading, you can set
    the convert keyword argument in spread to True.

    Args:
        key (str, int, or symbolic): Label for the key column.
        values (str, int, or symbolic): Label for the values column.

    Kwargs:
        convert (bool): Boolean indicating whether or not to try and convert
            the spread columns to more appropriate data types.


    Example:
        widened = elongated >> spread(X.variable, X.value)
        widened >> head(5)

            _ID carat clarity color        cut depth price table     x     y     z
        0     0  0.23     SI2     E      Ideal  61.5   326    55  3.95  3.98  2.43
        1     1  0.21     SI1     E    Premium  59.8   326    61  3.89  3.84  2.31
        2    10   0.3     SI1     J       Good    64   339    55  4.25  4.28  2.73
        3   100  0.75     SI1     D  Very Good  63.2  2760    56   5.8  5.75  3.65
        4  1000  0.75     SI1     D      Ideal  62.3  2898    55  5.83   5.8  3.62
    """

    # Taken mostly from dplython package
    columns = df.columns.tolist()
    id_cols = [col for col in columns if not col in [key, values]]

    temp_index = ['' for i in range(len(df))]
    for id_col in id_cols:
        temp_index += df[id_col].map(str)

    out_df = df.assign(temp_index=temp_index)
    out_df = out_df.set_index('temp_index')
    spread_data = out_df[[key, values]]

    if not all(spread_data.groupby([spread_data.index, key]).agg(
            'count').reset_index()[values] < 2):
        raise ValueError('Duplicate identifiers')

    spread_data = spread_data.pivot(columns=key, values=values)

    if convert and (out_df[values].dtype.kind in 'OSaU'):
        columns_to_convert = [col for col in spread_data if col not in columns]
        spread_data = convert_type(spread_data, columns_to_convert)

    out_df = out_df[id_cols].drop_duplicates()
    out_df = out_df.merge(spread_data, left_index=True, right_index=True).reset_index(drop=True)

    out_df = (out_df >> arrange(id_cols)).reset_index(drop=True)

    return out_df


# ------------------------------------------------------------------------------
# Separate columns
# ------------------------------------------------------------------------------

@pipe
@symbolic_evaluation(eval_as_label=['*'])
def separate(df, column, into, sep="[\W_]+", remove=True, convert=False,
             extra='drop', fill='right'):
    """
    Splits columns into multiple columns.

    Args:
        df (pandas.DataFrame): DataFrame passed in through the pipe.
        column (str, symbolic): Label of column to split.
        into (list): List of string names for new columns.

    Kwargs:
        sep (str or list): If a string, the regex string used to split the
            column. If a list, a list of integer positions to split strings
            on.
        remove (bool): Boolean indicating whether to remove the original column.
        convert (bool): Boolean indicating whether the new columns should be
            converted to the appropriate type.
        extra (str): either `'drop'`, where split pieces beyond the specified
            new columns are dropped, or `'merge'`, where the final split piece
            contains the remainder of the original column.
        fill (str): either `'right'`, where `np.nan` values are filled in the
            right-most columns for missing pieces, or `'left'` where `np.nan`
            values are filled in the left-most columns.
    """

    assert isinstance(into, (tuple, list))

    if isinstance(sep, (tuple, list)):
        inds = [0] + list(sep)
        if len(inds) > len(into):
            if extra == 'drop':
                inds = inds[:len(into) + 1]
            elif extra == 'merge':
                inds = inds[:len(into)] + [None]
        else:
            inds = inds + [None]

        splits = df[column].map(lambda x: [str(x)[slice(inds[i], inds[i + 1])]
                                           if i < len(inds) - 1 else np.nan
                                           for i in range(len(into))])

    else:
        maxsplit = len(into) - 1 if extra == 'merge' else 0
        splits = df[column].map(lambda x: re.split(sep, x, maxsplit))

    right_filler = lambda x: x + [np.nan for i in range(len(into) - len(x))]
    left_filler = lambda x: [np.nan for i in range(len(into) - len(x))] + x

    if fill == 'right':
        splits = [right_filler(x) for x in splits]
    elif fill == 'left':
        splits = [left_filler(x) for x in splits]

    for i, split_col in enumerate(into):
        df[split_col] = [x[i] if not x[i] == '' else np.nan for x in splits]

    if convert:
        df = convert_type(df, into)

    if remove:
        df.drop(column, axis=1, inplace=True)

    return df


# ------------------------------------------------------------------------------
# Unite columns
# ------------------------------------------------------------------------------

@pipe
@symbolic_evaluation(eval_as_label=['*'])
def unite(df, colname, *args, **kwargs):
    """
    Does the inverse of `separate`, joining columns together by a specified
    separator.

    Any columns that are not strings will be converted to strings.

    Args:
        df (pandas.DataFrame): DataFrame passed in through the pipe.
        colname (str): the name of the new joined column.
        *args: list of columns to be joined, which can be strings, symbolic, or
            integer positions.

    Kwargs:
        sep (str): the string separator to join the columns with.
        remove (bool): Boolean indicating whether or not to remove the
            original columns.
        na_action (str): can be one of `'maintain'` (the default),
            '`ignore'`, or `'as_string'`. The default will make the new column
            row a `NaN` value if any of the original column cells at that
            row contained `NaN`. '`ignore'` will treat any `NaN` value as an
            empty string during joining. `'as_string'` will convert any `NaN`
            value to the string `'nan'` prior to joining.
    """

    to_unite = list([a for a in flatten(args)])
    sep = kwargs.get('sep', '_')
    remove = kwargs.get('remove', True)
    # possible na_action values
    # ignore: empty string
    # maintain: keep as np.nan (default)
    # as_string: becomes string 'nan'
    na_action = kwargs.get('na_action', 'maintain')

    print(to_unite, sep, remove, na_action)

    if na_action == 'maintain':
        df[colname] = df[to_unite].apply(lambda x: np.nan if any(x.isnull())
        else sep.join(x.map(str)), axis=1)
    elif na_action == 'ignore':
        df[colname] = df[to_unite].apply(lambda x: sep.join(x[~x.isnull()].map(str)),
                                         axis=1)
    elif na_action == 'as_string':
        df[colname] = df[to_unite].astype(str).apply(lambda x: sep.join(x), axis=1)

    if remove:
        df.drop(to_unite, axis=1, inplace=True)

    return df
